File: drills/ts/solutions/nc73_word_break.ts
export function wordBreak(s: string, wordDict: string[]): boolean {
  const words = new Set(wordDict);

  // Reachable positions rather than a table of booleans: start at 0, and a
  // position is reachable when some word in the dictionary bridges the gap from
  // a position already reached.
  const reached = new Set([0]);
  for (let end = 1; end <= s.length; end++) {
    for (let start = 0; start < end; start++) {
      if (reached.has(start) && words.has(s.slice(start, end))) {
        reached.add(end);
        break;
      }
    }
  }

  return reached.has(s.length);
}

File: drills/go/solutions/nc73_word_break.go
package main

func wordBreak(s string, wordDict []string) bool {
	words := map[string]bool{}
	for _, w := range wordDict {
		words[w] = true
	}
	// breakable[i]: can s[:i] be segmented? True when some word ends at i
	// and the prefix before it was breakable.
	breakable := make([]bool, len(s)+1)
	breakable[0] = true
	for end := 1; end <= len(s); end++ {
		for start := 0; start < end; start++ {
			if breakable[start] && words[s[start:end]] {
				breakable[end] = true
				break
			}
		}
	}
	return breakable[len(s)]
}
